persistence/pg.go

```diff
@@ -249,53 +251,77 @@ func (p *PostgresPersister) getHtlcHash(ctx context.Context, tx *pg.Tx,
 	return htlc.Hash, nil
 }
 
-func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
-	key types.HtlcKey) (*lntypes.Hash, error) {
+func (p *PostgresPersister) MarkHtlcFinal(ctx context.Context,
+	key types.HtlcKey, settled bool) (*lntypes.Hash, error) {
 
-	var settledHash *lntypes.Hash
+	var invoiceHash *lntypes.Hash
+
+	var finalStatus types.Status
+	if settled {
+		finalStatus = types.StatusSettled
+	} else {
+		finalStatus = types.StatusFailed
+	}
 
 	err := p.conn.RunInTransaction(ctx, func(tx *pg.Tx) error {
-		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
-		// htlcs for the same invoice are marked as settled concurrently.
+		// Select invoice FOR UPDATE to prevent concurrent update on the
+		// invoice/htlc.
 		invoice, err := p.selectInvoiceForUpdate(ctx, tx, key)
 		if err != nil {
 			return err
 		}
 
 		now := time.Now().UTC()
 
-		// Settle the HTLC in the database.
-		if err := p.settleHTLC(ctx, tx, key, now); err != nil {
-			return err
-		}
-
-		// The HTLC is settled, now we check if all other HTLCs
-		// are settled as well.
-		allHtlcsSettled, err := p.allHtlcsAreSettled(ctx,
-			tx, invoice.Hash)
+		err = p.changeHTLCStatus(ctx, tx, key, finalStatus, now)
 		if err != nil {
 			return err
 		}
-		if !allHtlcsSettled {
+
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have reached
+		// its final status afterward.
+		if invoice.Status == types.StatusFailed {
```

Comment:
In which case this will happen?

I mean, if invoice has StatusSettled state, it means that all HTLC were settled and MarkHtlcFinal will never be called again for this invoice